addons/mail/static/src/utils/common/format.js

```diff
@@ -179,15 +179,16 @@ function generateMentionsLinks(body, { partners = [], threads = [], specialMenti
     }
     for (const thread of threads) {
         const placeholder = `#-mention-channel-${thread.id}`;
-        const text = `#${escape(thread.displayName)}`;
+        const iconClass = thread.parent_channel_id ? "fa-comments-o" : "fa-hashtag";
+        const text = `<i class="fa ${iconClass}"></i>${escape(thread.displayName)}`;
         mentions.push({
             class: "o_channel_redirect",
             id: thread.id,
             model: "discuss.channel",
             placeholder,
             text,
         });
-        body = body.replace(text, placeholder);
+        body = body.replace(`#${escape(thread.displayName)}`, placeholder);
     }
```

Comment:
Diff here should only remove `#`, the message content should not have `<i class="fa">`.

Reason is that `<i class="fa">` on the message makes sense with current style sheets and FA version.
Migration scripts on older messages is a no-no, so the better tradeoff is to store pure logic inside the message.

Ideally only id of channel matters, the name of conversation in `text` shouldn't be there. But let's keep it for now because this is a more important change and it's not a big deal to maintain it.

For the styling however, the solution should add `<i class="fa">` on `.o_channel_redirect` in post-process of message body in JS.

So you should make sure to not send the icon while posting or editing a new message, and the display on UI should have the icon. The icon is a decoration made solely in JS.